File: scripts/seed.ts
/**
 * One-shot seed for PSU campus/college/program rows.
 * Run with:  npm run seed
 *
 * Requires NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY in env.
 */
import { readFile } from "node:fs/promises";
import path from "node:path";
import { createClient } from "@supabase/supabase-js";
import { parseCourses } from "../lib/parsers";

async function main() {
  const url = process.env.NEXT_PUBLIC_SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!url || !key) {
    console.error("Set NEXT_PUBLIC_SUPABASE_URL and SUPABASE_SERVICE_ROLE_KEY first.");
    process.exit(1);
  }
  const db = createClient(url, key, { auth: { persistSession: false } });

  const csvPath = path.join(process.cwd(), "data", "psu_programs.csv");
  const buf = await readFile(csvPath);
  const records = await parseCourses("psu_programs.csv", buf);

  const rows = records.map((r) => ({
    campus: r.campus ?? "",
    college: r.college ?? "",
    program: r.program ?? "",
    major: r.major ?? "",
    course_code: "",
    course_title: r.major ? `${r.program} - ${r.major}` : r.program ?? "Untitled",
    description: `${r.program ?? ""} program at ${r.college ?? ""}, ${r.campus ?? ""}.${r.major ? ` Major/track: ${r.major}.` : ""}`,
    learning_outcomes: "",
    keywords: r.major ?? "",
    enrollment: 0,
  }));

  let inserted = 0;
  for (let i = 0; i < rows.length; i += 500) {
    const slice = rows.slice(i, i + 500);
    const { data, error } = await db.from("courses").insert(slice).select("id");
    if (error) { console.error(error); process.exit(2); }
    inserted += data?.length ?? 0;
  }
  console.log(`Inserted ${inserted} program rows.`);
}

main();

File: app/seed.py
"""Seed PSU campus/college/program data from data/psu_programs.csv."""
from pathlib import Path
import pandas as pd

from .db import SessionLocal, init_db
from . import models

CSV_PATH = Path(__file__).resolve().parent.parent / "data" / "psu_programs.csv"


def seed_programs() -> int:
    """Insert PSU programs as Course rows (one per Program+Major)."""
    init_db()
    if not CSV_PATH.exists():
        return 0
    df = pd.read_csv(CSV_PATH).fillna("")
    df.columns = [c.strip() for c in df.columns]

    inserted = 0
    with SessionLocal() as db:
        existing = {
            (c.campus, c.college, c.program, c.major)
            for c in db.query(models.Course).all()
        }
        for _, row in df.iterrows():
            campus = str(row.get("Campus", "")).strip()
            college = str(row.get("College", "")).strip()
            program = str(row.get("Program / Degree", "")).strip()
            major = str(row.get("Major/ Track / Specialization", "")).strip()
            if not program:
                continue
            key = (campus, college, program, major)
            if key in existing:
                continue
            course = models.Course(
                campus=campus,
                college=college,
                program=program,
                major=major,
                course_code="",
                course_title=program + (f" - {major}" if major else ""),
                description=f"{program} program at {college}, {campus}." + (f" Major/track: {major}." if major else ""),
                learning_outcomes="",
                keywords=major,
                enrollment=0,
            )
            db.add(course)
            existing.add(key)
            inserted += 1
        db.commit()
    return inserted


if __name__ == "__main__":
    n = seed_programs()
    print(f"Seeded {n} program rows.")
